sklearn/neural_network/multilayer_perceptron.py

```diff
@@ -0,0 +1,1248 @@
+                 validation_fraction, beta_1, beta_2, epsilon):
+        self.activation = activation
+        self.algorithm = algorithm
+        self.alpha = alpha
+        self.batch_size = batch_size
+        self.learning_rate = learning_rate
+        self.learning_rate_init = learning_rate_init
+        self.power_t = power_t
+        self.max_iter = max_iter
+        self.loss = loss
+        self.hidden_layer_sizes = hidden_layer_sizes
+        self.shuffle = shuffle
+        self.random_state = random_state
+        self.tol = tol
+        self.verbose = verbose
+        self.warm_start = warm_start
+        self.momentum = momentum
+        self.nesterovs_momentum = nesterovs_momentum
+        self.early_stopping = early_stopping
+        self.validation_fraction = validation_fraction
+        self.beta_1 = beta_1
+        self.beta_2 = beta_2
+        self.epsilon = epsilon
+
+    def _unpack(self, packed_parameters):
+        """Extract the coefficients and intercepts from packed_parameters."""
+        for i in range(self.n_layers_ - 1):
+            start, end, shape = self._coef_indptr[i]
+            self.coefs_[i] = np.reshape(packed_parameters[start:end], shape)
+
+            start, end = self._intercept_indptr[i]
+            self.intercepts_[i] = packed_parameters[start:end]
+
+    def _forward_pass(self, activations, with_output_activation=True):
+        """Perform a forward pass on the network by computing the values
+        of the neurons in the hidden layers and the output layer.
+
+        Parameters
+        ----------
+        activations: list, length = n_layers - 1
+            The ith index of the list holds the values of the ith layer.
+
+        with_output_activation : bool, default True
+            If True, the output passes through the output activation
+            function, which is either the softmax function or the
+            logistic function
+        """
+        # Iterate over the hidden layers
+        for i in range(self.n_layers_ - 1):
+            activations[i + 1] = safe_sparse_dot(activations[i],
+                                                 self.coefs_[i])
+            activations[i + 1] += self.intercepts_[i]
+
+            # For the hidden layers
+            if i + 1 != self.n_layers_ - 1:
+                hidden_activation = ACTIVATIONS[self.activation]
+                activations[i + 1] = hidden_activation(activations[i + 1])
+
+        # For the last layer
+        if with_output_activation:
+            output_activation = ACTIVATIONS[self.out_activation_]
+            activations[i + 1] = output_activation(activations[i + 1])
+
+        return activations
+
+    def _compute_loss_grad(self, layer, n_samples, activations, deltas,
+                           coef_grads, intercept_grads):
+        """Compute the loss gradient for the layer."""
+        coef_grads[layer] = safe_sparse_dot(activations[layer].T,
```

Comment:
I think you are right, since all weights are dense, only the input can be sparse. All intermediate layers should be dense. I'm also ok with either having `safe_sparse_dot` to be safe or change them to dense version of dot.